pkg/comp-functions/functions/vshnredis/mailgun_alerting.go

```diff
@@ -0,0 +1,147 @@
+package vshnredis
+
+import (
+	"context"
+	"strconv"
+
+	xkube "github.com/crossplane-contrib/provider-kubernetes/apis/object/v1alpha1"
+	alertmanagerv1alpha1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1alpha1"
+	vshnv1 "github.com/vshn/appcat/v4/apis/vshn/v1"
+	runtime "github.com/vshn/appcat/v4/pkg/comp-functions/runtime"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/pointer"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+)
+
+func MailgunAlerting(ctx context.Context, iof *runtime.Runtime) runtime.Result {
+	log := controllerruntime.LoggerFrom(ctx)
+	log.Info("Starting mailgun-alerting function")
+
+	comp := &vshnv1.VSHNRedis{}
+	err := iof.Observed.GetComposite(ctx, comp)
+	if err != nil {
+		return runtime.NewFatalErr(ctx, "Cannot get composite from function io", err)
+	}
+
+	if !mailAlertingEnabled(iof.Config) {
+		if comp.Spec.Parameters.Monitoring.Email == "" {
+			return runtime.NewNormal()
+		}
+		return runtime.NewWarning(ctx, "Email Alerting is not enabled")
+	}
+	// Wait for the next reconciliation in case instance namespace is missing
+	if comp.Status.InstanceNamespace == "" {
+		return runtime.NewWarning(ctx, "Composite is missing instance namespace, skipping transformation")
+	}
+
+	if comp.Spec.Parameters.Monitoring.Email == "" {
+		return runtime.NewNormal()
+	}
+
+	log.Info("Deploying AlertmanagerConfig for mail alerting...")
+	err = deployAlertmanagerConfig(ctx, comp, iof)
+	if err != nil {
+		return runtime.NewFatalErr(ctx, "Can't deploy AlertmanagerConfig "+comp.Name+"-alertmanagerconfig-mailgun for mail alerting", err)
+	}
+
+	log.Info("Finishing mailgun-alerting function with NewNormal")
+
+	return runtime.NewNormal()
+}
+
+func deployAlertmanagerConfig(ctx context.Context, comp *vshnv1.VSHNRedis, iof *runtime.Runtime) error {
```

Comment:
Same as above:
Can't we work with generics here and put this function in a common place and reuse it for both services?
It's basically still a copy&paste from the postgreSQL FunctionIO

Reply:
went with interface solution, generics for this use case caused me a massive headache because of their limitations

Reply:
But we do we still have this function here? you have a function of the same name in the common package, with a different set of arguments however.

Reply:
the common one is the good one, I left old funcs in postgresql and redis directories